File: Ruby/pig.rb
require 'sinatra'
require 'pg'
require 'json'

# database connection, to be used in request handling
DB = PG::Connection.new(dbname: 'pig', user: 'pig')

# read the database schema, to reset test data
DIR = File.expand_path('../..', __FILE__)
SQL = File.read(DIR + '/pig.sql')

error Sinatra::NotFound do
  content_type 'application/json'
  [404, '{}']
end

class Pig
  attr_accessor :res

  # prefix every function call with this schema name
  # (database has other schemas. we just one this one)
  def initialize(schema)
    @schema = schema
  end

  # make the "($1,$2)" string for inside exec_params query
  def paramstring(params)
    '(%s)' % (1..params.size).map {|i| "$#{i}"}.join(',')
  end

  # INPUT: function name, varargs parameters
  # OUTPUT: @res['status'] (HTTP status), @res['js'] (JSON)
  def q(func, *params)
    @res = DB.exec_params("SELECT status, js FROM %s.%s%s" %
      [@schema, func, paramstring(params)], params)[0]
  end
end

# before processing every HTTP request...
before do
  # get the apikey from HTTP header. quick fail if none or malformed
  apikey = request.env['HTTP_APIKEY']
  if apikey.nil? || !(/\A[a-z]{4}\Z/ === apikey)
    halt 401, '{"error":"needs apikey header"}'
  end

  # reset the database test info
  DB.exec(SQL)

  # init database connection for use in all handlers
  # 'pig' means every function will be prefixed with that schema name
  @pig = Pig.new('pig')

  # lookup person_id using apikey. remember it if found. fail now if not.
  @pig.q('apikey_get', apikey)
  if '200' == @pig.res['status']
    @person_id = JSON.parse(@pig.res['js'])['person_id']
    @pig.res = nil
  else
    @pig = nil
    halt 401, '{"error":"wrong apikey"}'
  end
end

# nice shortcut instead of manually passing status and js to response
# with each request, the "after" filter does it every time
# the "if" is only because of the halts in the "before" filter
after do
  content_type 'application/json'
  if @pig && @pig.res
    status @pig.res['status']
    body @pig.res['js']
  end
end

# URI route handlers should be self-explanatory
# Note that @person_id from auth stage is used, to limit user to
# only get/set/update/delete their own things

get '/' do
  @pig.q('people_get')
end

get %r{/person/([1-9][0-9]{0,5})} do |id|
  @pig.q('person_get', id)
end

patch '/person' do
  halt(412, '{"error":"missing name"}') unless params[:name]
  @pig.q('person_update', @person_id, params[:name])
end

get '/things' do
  @pig.q('things_get', @person_id)
end

get %r{/thing/([1-9][0-9]{0,5})} do |id|
  @pig.q('thing_get', @person_id, id)
end

patch %r{/thing/([1-9][0-9]{0,5})} do |id|
  halt(412, '{"error":"missing name"}') unless params[:name]
  @pig.q('thing_update', @person_id, id, params[:name])
end

post '/things' do
  halt(412, '{"error":"missing name"}') unless params[:name]
  @pig.q('thing_add', @person_id, params[:name])
end

delete %r{/thing/([1-9][0-9]{0,5})} do |id|
  @pig.q('thing_delete', @person_id, id)
end

File: Python/pig.py
import psycopg2
from flask import Flask, jsonify, request, abort, Response
import os
import psycopg2.extras
import re
from functools import wraps
from werkzeug.routing import BaseConverter

app = Flask(__name__)


class RegexConverter(BaseConverter):
    def __init__(self, url_map, *items):
        super(RegexConverter, self).__init__(url_map)
        self.regex = items[0]


app.url_map.converters["regex"] = RegexConverter
conn = psycopg2.connect("dbname=pig user=pig")
conn.autocommit = True
DB = conn.cursor(cursor_factory=psycopg2.extras.DictCursor)

DIR = os.path.abspath(".")
with open(DIR + "/pig.sql") as f:
    SQL = f.read()


@app.errorhandler(404)
def not_found(error):
    return jsonify("{}"), 404


class MissingApikey(Exception):
    pass


@app.errorhandler(MissingApikey)
def missing_apikey(error):
    return jsonify("{'error': 'needs apikey header'}"), 401


class WrongApikey(Exception):
    pass


@app.errorhandler(WrongApikey)
def wrong_apikey(error):
    return jsonify("{'error':'wrong apikey'}"), 401

class MissingName(Exception):
    pass

@app.errorhandler(MissingName)
def missing_name(error):
    return jsonify("{'error':'missing name'}"), 412


class Pig:
    def __init__(self, schema):
        self.schema = schema

    def format_parameter(self, num):
        return f"%s"

    def paramstring(self, num):
        list_of_nums = list(range(1, num + 1))
        joined_nums = map(self.format_parameter, list_of_nums)
        return f"({','.join(joined_nums)})"

    def q(self, func, *params):
        DB.execute(
            f"SELECT status, js FROM {self.schema}.{func}{self.paramstring(len(params))}",
            params,
        )
        self.res = DB.fetchone()


def before():
    apikey = request.headers.get("Apikey")
    if (apikey is None) or (re.search("\A[a-z]{4}\Z", apikey) is None):
        raise MissingApikey

    DB.execute(SQL)

    pig_ = Pig("pig")

    pig_.q("apikey_get", apikey)
    print(pig_.res)
    if 200 == pig_.res["status"]:
        pig_.person_id = pig_.res["js"].get("person_id")
        pig_.res = None
    else:
        pig_.res = None
        raise WrongApikey

    return pig_


def after(pig_):
    if pig_ and pig_.res:
        return jsonify(pig_.res["js"]), pig_.res["status"]


def before_and_after():
    def decorator_func(func):
        @wraps(func)
        def wrapper_func(*args, **kwargs):
            pig_ = before()
            retval = func(pig_, *args, **kwargs)
            return after(pig_)

        return wrapper_func

    return decorator_func


@app.route("/")
@before_and_after()
def people_get(pig_):
    pig_.q("people_get")


@app.route("/person/<regex('[1-9][0-9]{0,5}'):id>")
@before_and_after()
def person_get(pig_, id):
    pig_.q("person_get", id)

@app.route("/person", methods=["PATCH"])
@before_and_after()
def person_update(pig_):
    if request.form.get("name") is None:
        raise MissingName    
    pig_.q("person_update", pig_.person_id, request.form.get("name"))


if __name__ == "__main__":
    app.run(debug=True)
